evaluate/report/csv.go

```diff
@@ -26,50 +27,109 @@ type CSVFormatter interface {
 	Rows() (rows [][]string)
 }
 
-// GenerateCSV returns the whole CSV as string.
-func GenerateCSV(formatter CSVFormatter) (csvData string, err error) {
-	var out strings.Builder
-	csv := csv.NewWriter(&out)
+// EvaluationRecord holds a line of the evaluation CSV.
+type EvaluationRecord struct {
+	// ModelID holds the model id.
+	ModelID string
+	// ModelName holds the model name.
+	ModelName string
+	// ModelCost holds the model cost.
+	ModelCost float64
 
-	if err := csv.Write(formatter.Header()); err != nil {
-		return "", pkgerrors.WithStack(err)
-	}
+	// LanguageID holds the language id.
+	LanguageID string
 
-	for _, row := range formatter.Rows() {
-		if err := csv.Write(row); err != nil {
-			return "", pkgerrors.WithStack(err)
+	// Assessments holds the assessments of an entry.
+	Assessments metrics.Assessments
+}
+
+// Clone clones an evaluation record.
+func (e *EvaluationRecord) Clone() (new *EvaluationRecord) {
+	new = &EvaluationRecord{}
+
+	new.ModelID = e.ModelID
+	new.ModelName = e.ModelName
+	new.ModelCost = e.ModelCost
+	new.LanguageID = e.LanguageID
+	new.Assessments = metrics.Merge(e.Assessments, nil)
+
+	return new
+}
+
+// EvaluationRecords holds all the evaluation records.
+type EvaluationRecords []*EvaluationRecord
+
+// EvaluationRecordsPerModel holds the collection of evaluation records per model.
+type EvaluationRecordsPerModel map[string]*EvaluationRecord
+
+// GroupByModel groups the evaluation records by model.
+func (e EvaluationRecords) GroupByModel() EvaluationRecordsPerModel {
+	perModel := map[string]*EvaluationRecord{}
+
+	for _, record := range e {
+		_, ok := perModel[record.ModelID]
+		if !ok {
+			perModel[record.ModelID] = record.Clone()
+		} else {
+			r := perModel[record.ModelID]
+			r.Assessments = metrics.Merge(r.Assessments, record.Assessments)
 		}
 	}
 
-	csv.Flush()
-
-	return out.String(), nil
+	return perModel
```

Comment:
Since you moved some grouping functionality here, I would think that the assessment package now has some functions that are unused, right? Can you please check? I would think that the only reason we have all the assessment magic is the categorization. Which we should either remove completely or move to here at some point.

Reply:
Removed the unused functions/structs. There are still some things used in the markdown report and in the evaluation itself.

Reply:
Cannot wait to get rid of all of that as well.